mzLib/Proteomics/RetentionTimePrediction/ChronologerModel/ChronologerEstimator.cs

```diff
@@ -0,0 +1,344 @@
+        /// <summary>
+        /// Uses the Chronologer model to predict C18 retention times (reported in % ACN).
+        /// Only modifications present in the Chronologer dictionary are supported.
+        /// Returns null if the sequence is not valid.
+        /// <code>
+        /// "Carbamidomethyl on C"
+        /// "Oxidation on M"
+        /// "Glu to PyroGlu"
+        /// "Phosphorylation on S"
+        /// "Phosphorylation on T"
+        /// "Phosphorylation on Y"
+        /// "Acetylation on K"
+        /// "Succinylation on K"
+        /// "Ubiquitination on K"
+        /// "Methylation on K"
+        /// "Dimethylation on K"
+        /// "Trimethylation on K"
+        /// "Methylation on R"
+        /// "Dimethylation on R"
+        /// </code>
+        /// </summary>
+        /// <param name="baseSequence"></param>
+        /// <param name="fullSequence"></param>
+        /// <returns></returns>
+        public static float PredictRetentionTime(string baseSequence, string fullSequence)
+        {
+            if (baseSequence.Length > 50)
+            {
+                return -1;
+            }
+            var tensor = Tensorize(baseSequence, fullSequence);
+
+            TorchSharp.Utils.TensorAccessor<long> tensorAccessor = tensor.data<long>();
+            long[] tensorAsArray = tensorAccessor.ToArray();
+            long[] incompatibleTensor = new long[52];
+            incompatibleTensor.Fill(0);
+
+            if (tensorAsArray.SequenceEqual(incompatibleTensor))
+            {
+                return -1;
+            }
+
+            torch.Tensor prediction = ChronologerModel.Predict(tensor);
+            return prediction.data<float>()[0];
+        }
+
+        public static float[] PredictRetentionTime(string[] baseSequences, string[] fullSequences, bool gpu = true, int batch_size = 512)
+        {
+            torch.Device device;
+            try
+            {
+                torch.InitializeDeviceType(DeviceType.CUDA);
+                device = new torch.Device(DeviceType.CUDA);
+            }
+            catch
+            {
+                Console.WriteLine(new Exception("This device does not support CUDA, using CPU"));
+                device = new torch.Device(DeviceType.CPU);
+            }
+
+            ChronologerModel.to(device);
+
+            //vstack tensors and then batch predict to avoid running out of vRAM, I think system RAM transfer has big latency so keep em small
+
+            float[] predictions = new float[baseSequences.Length];
+
+            // tensorize all
+            torch.Tensor[] tensorsArray = new torch.Tensor[baseSequences.Length];
+            bool[] compatibleTracker = new bool[baseSequences.Length];
+
+            Parallel.For(0, baseSequences.Length, (i, state) =>
+            {
+                var baseSeq = baseSequences[i];
+                var fullSeq = fullSequences[i];
+
+                var (tensor, compatible) = Tensorize(baseSeq, fullSeq, out bool chronologerCompatible);
```

Comment:
what is the difference between compatible and chronologerCompatible? Why is one checked but not the other?